File: Contents/OutLineExpandable/Outline Project/MainTableViewController.m
//
//  MainTableViewController.m
//  MGRLatex
//
//  Created by Kwan Hyun Son on 2021/07/20.
//

#import "MainTableViewController.h"
#import "ItemsForTableView.h"
#import "Item.h"
#import "ViewController.h"
#import "ViewControllerX1.h"
#import "ViewControllerX2.h"
#import "ViewControllerY1.h"
#import "ViewControllerY2.h"

@interface MainTableViewController ()

@end

@implementation MainTableViewController

- (instancetype)initWithStyle:(UITableViewStyle)style {
    if (@available(iOS 13, *)) {
        self = [super initWithStyle:UITableViewStyleInsetGrouped];
    } else {
        self = [super initWithStyle:UITableViewStyleGrouped];
    }
    
    if(self) {
        self.navigationItem.title = @"Outline Project";
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
//    [self.tableView registerClass:[UITableViewCell class]
//           forCellReuseIdentifier:@"UITableViewCell"];
}

#pragma mark - <UITableViewDataSource>
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return ItemsForTableView.sharedItems.allItems.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return ItemsForTableView.sharedItems.allItems[section].allValues.firstObject.count;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    NSMutableDictionary <NSString *, NSArray <Item *>*>*itemsDic =
    ItemsForTableView.sharedItems.allItems[section];
    if (section == 0) {
        return itemsDic.allKeys.firstObject;
    } else if (section == 1) {
        return itemsDic.allKeys.firstObject;
    } else {
        return itemsDic.allKeys.firstObject;
    }

    return @"뉘미럴";
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"jotto"];
    if (cell == nil) {
//        NSLog(@"처음만들어");
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"jotto"];
    } else {
//        NSLog(@"재사용이야");
    }
    
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator; /// 오른쪽에 > 표시를 만든다.
    
    NSMutableDictionary <NSString *, NSArray <Item *>*>*itemsDic =
    ItemsForTableView.sharedItems.allItems[indexPath.section];
    Item *item = itemsDic.allValues.firstObject[indexPath.row];
    
    if (@available(iOS 14, *)) {
        UIListContentConfiguration *content = [cell defaultContentConfiguration];
        content.text = item.textLabelText;
        content.secondaryText = item.detailTextLabelText;

// 이거 쓰면 방향 Portrait <=> Landscape 일때, 이상 발생한다.
//        NSDirectionalEdgeInsets directionalEdgeInsets = cell.directionalLayoutMargins;
//        directionalEdgeInsets.top = directionalEdgeInsets.top + 1.0;
//        directionalEdgeInsets.bottom = directionalEdgeInsets.bottom + 1.0;
        content.directionalLayoutMargins = NSDirectionalEdgeInsetsMake(8.0, 8.0, 8.0, 8.0);
        content.textToSecondaryTextVerticalPadding = 5.0;
        cell.contentConfiguration = content;
    } else {
        cell.textLabel.text = item.textLabelText;
        cell.detailTextLabel.text = item.detailTextLabelText;
    }
    
    return cell;
}

#pragma mark - <UITableViewDelegate>
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
    UIListContentConfiguration *listContentConfiguration = (UIListContentConfiguration *)cell.contentConfiguration;
    UIViewController *viewController;
    
    if (indexPath.section == 0) {
        if(indexPath.row == 0){
            viewController = [ViewController new];
        } 
        if (@available(iOS 14, *)) {
            viewController.navigationItem.title = listContentConfiguration.text;
        } else {
            viewController.navigationItem.title = cell.detailTextLabel.text;
        }
    }
    
    if (indexPath.section == 1) {
        if(indexPath.row == 0){
            viewController = [ViewControllerX1 new];
        } else if (indexPath.row == 1) {
            viewController = [ViewControllerX2 new];
        }
        if (@available(iOS 14, *)) {
            viewController.navigationItem.title = [NSString stringWithFormat:@"%@ %@", listContentConfiguration.text, listContentConfiguration.secondaryText];
        } else {
            viewController.navigationItem.title = [NSString stringWithFormat:@"%@ %@", cell.textLabel.text, cell.detailTextLabel.text];
        }
    }
    
    if (indexPath.section == 2) {
        if(indexPath.row == 0){
            viewController = [ViewControllerY1 new];
        } else if(indexPath.row == 1){
            viewController = [ViewControllerY2 new];
        }
        
        if (@available(iOS 14, *)) {
            viewController.navigationItem.title = [NSString stringWithFormat:@"%@ %@", listContentConfiguration.text, listContentConfiguration.secondaryText];
        } else {
            viewController.navigationItem.title = [NSString stringWithFormat:@"%@ %@", cell.textLabel.text, cell.detailTextLabel.text];
        }
    }
    
    if (viewController != nil) {
        [self.navigationController pushViewController:viewController animated:YES];
    }
}

@end

File: Contents/FinancialKeyboard/IosSwiftSample/MainTableViewController.swift
//
//  ViewController.swift
//  MGUNeoSegControl
//
//  Created by Kwan Hyun Son on 2022/10/08.
//

import UIKit
import BaseKit
import IosKit

final class TableViewDiffableDataSource: UITableViewDiffableDataSource<String, Item> {
    // MARK: header/footer titles support
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let allItems = ItemsForTableView.shared.allItems
        return allItems[section].sectionTitle
    }
}

final class MainTableViewController: UIViewController {
    
    private var dataSource : TableViewDiffableDataSource?
    private var currentSnapshot: NSDiffableDataSourceSnapshot<String, Item>?
    
    private lazy var tableView = {
        return UITableView(frame: .zero, style: .insetGrouped)
    }()
    
    private var iOSBundle: Bundle? = {
        return Bundle.skhIosRes
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "Financial Keyboard"
        configureTableView()
        configureDataSource()
        updateUI(animated: false)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        let indexPath = tableView.indexPathForSelectedRow
        if let indexPath = indexPath {
            tableView.deselectRow(at: indexPath, animated: true)
        }
    }

    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
}

extension MainTableViewController {
    func configureTableView() {
        tableView = UITableView(frame: .zero, style: .insetGrouped)
        tableView.rowHeight = UITableView.automaticDimension
        tableView.estimatedRowHeight = 50.0
        view.addSubview(tableView)
        tableView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            tableView.topAnchor.constraint(equalTo: view.topAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: String(describing: UITableViewCell.self))
        tableView.delegate = self
    }
    
    func configureDataSource() {
        dataSource = TableViewDiffableDataSource(tableView: tableView) { (tableView, indexPath, item) -> UITableViewCell? in

            let cell = tableView.dequeueReusableCell(
                withIdentifier: String(describing: UITableViewCell.self),
                for: indexPath
            )

            if indexPath.section == 1 {
                cell.accessoryType = .disclosureIndicator
            }

            if #available(iOS 14.0, *) {
                var content = cell.defaultContentConfiguration()
                content.text = item.title
                content.secondaryText = item.detailText
                content.textProperties.color = .label
                // content.textProperties.font = .boldSystemFont(ofSize: 13.0)
                content.secondaryTextProperties.color = .secondaryLabel
                // content.secondaryTextProperties.font = .systemFont(ofSize: 12.0)
                content.directionalLayoutMargins = NSDirectionalEdgeInsets(top: 8.0, leading: 8.0, bottom: 8.0, trailing: 8.0)
                content.textToSecondaryTextVerticalPadding = 5.0
                cell.contentConfiguration = content
            } else {
                cell.textLabel?.text = item.title
                cell.detailTextLabel?.text = item.detailText
            }

            return cell
        }
        self.dataSource?.defaultRowAnimation = .fade
    }
    
    func updateUI(animated: Bool = true) {
        currentSnapshot = NSDiffableDataSourceSnapshot<String, Item>()
        let allItems = ItemsForTableView.shared.allItems
        for i in 0..<allItems.count {
            let section = allItems[i]
            let sectionTitle = section.sectionTitle
            let items = section.items
            
            currentSnapshot?.appendSections([sectionTitle])
            currentSnapshot?.appendItems(items, toSection:sectionTitle)
        }
        
        dataSource?.apply(currentSnapshot!, animatingDifferences: animated)
    }
}

// MARK: - UITableViewDelegate
extension MainTableViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        var viewController: UIViewController?
        if indexPath.section == 0 {
        viewController = ViewControllerA()
        } else if indexPath.section == 1 {
            if indexPath.row == 0 {
                let board = UIStoryboard(name: "ViewControllerB", bundle: nil)
                viewController = board.instantiateViewController(withIdentifier: "ViewControllerB")
            } else if indexPath.row == 1 {
                let board = UIStoryboard(name: "ViewControllerF", bundle: nil)
                viewController = board.instantiateViewController(withIdentifier: "ViewControllerF")
            }
        } else if indexPath.section == 2 {
            switch indexPath.row {
            ///case 0: viewController = ViewController3.init()
            ///case 1: viewController = ViewController4.init()
            default: break
            }
        }
        
        if let viewController = viewController {
            let allItems = ItemsForTableView.shared.allItems
            let section = allItems[indexPath.section]
            let title = section.items[indexPath.row].title
            viewController.navigationItem.title = title
            navigationController?.pushViewController(viewController, animated:true)
        }
    }
}
